include/revng/Storage/StorageClient.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+
+namespace revng {
+
+class WritableFile {
+private:
+  std::unique_ptr<llvm::raw_pwrite_stream> OS;
+  std::function<llvm::Error()> Closer;
+  std::function<void()> Aborter;
+
+public:
+  WritableFile(std::unique_ptr<llvm::raw_pwrite_stream> &&OS,
+               std::function<llvm::Error()> &&Closer,
+               std::function<void()> &&Aborter) :
+    OS(std::move(OS)), Closer(std::move(Closer)), Aborter(std::move(Aborter)) {}
+  WritableFile(const WritableFile &Other) = delete;
+  WritableFile &operator=(const WritableFile &Other) = delete;
+  WritableFile(WritableFile &&Other) = default;
+  WritableFile &operator=(WritableFile &&Other) = default;
+  ~WritableFile() {
+    if (OS != nullptr)
+      Aborter();
+  }
+
+  llvm::raw_pwrite_stream &getOS() { return *OS; }
+
+  llvm::Error close() {
+    revng_assert(OS);
+    OS.reset();
+    return Closer();
+  }
+};
+
+class ReadableFile {
+private:
+  std::unique_ptr<llvm::MemoryBuffer> Buffer;
+  std::function<void()> Closer;
+
+public:
+  ReadableFile(std::unique_ptr<llvm::MemoryBuffer> &&Buffer,
+               std::function<void()> &&Closer) :
+    Buffer(std::move(Buffer)), Closer(std::move(Closer)) {
+    revng_assert(this->Buffer.get() != nullptr);
+  }
+  ReadableFile(const ReadableFile &Other) = delete;
+  ReadableFile &operator=(const ReadableFile &Other) = delete;
+  ReadableFile(ReadableFile &&Other) = default;
+  ReadableFile &operator=(ReadableFile &&Other) = default;
+  ~ReadableFile() {
+    if (Buffer != nullptr)
+      Closer();
+  }
+
+  llvm::MemoryBuffer &getBuffer() { return *Buffer; }
+};
+
+enum ContentEncoding {
+  None,
+  Gzip
+};
+
+class StorageClient {
+protected:
+  StorageClient() = default;
+
+public:
+  virtual ~StorageClient() = default;
+  StorageClient(const StorageClient &Other) = delete;
+  StorageClient &operator=(const StorageClient &Other) = delete;
+  StorageClient(const StorageClient &&Other) = delete;
+  StorageClient &operator=(StorageClient &&Other) = delete;
+
+public:
+  static llvm::Expected<std::unique_ptr<StorageClient>>
+  fromURL(const llvm::StringRef URL);
+
+  virtual std::string toString() const debug_function = 0;
```

Comment:
What does `toString` return? The root URL from which we can rebuild this object using `fromURL`?

If not, it it's just for debug purpose, use `dump` and write to `dbg`.

Reply:
Renamed to `dump` and added the `dump(T &OS)` method as well (as specified in the pad)